File: src/main/java/dev/klomptech/jbark/interpreter/BarkValue.java
package dev.klomptech.jbark.interpreter;

public sealed interface BarkValue
    permits BarkValue.BarkString, BarkValue.BarkNumber, BarkValue.BarkBoolean, BarkValue.BarkNull {

  // Human readable representation of the value
  String display();

  // For error messages when value may be null
  static String describe(final BarkValue value) {
    return value == null ? "nothing" : value.display();
  }

  // Same type and same data?
  boolean equalTo(BarkValue other);

  record BarkString(String value) implements BarkValue {
    @Override
    public String display() {
      return value;
    }

    @Override
    public boolean equalTo(final BarkValue other) {
      return other instanceof BarkString(String v) && value.equals(v);
    }
  }

  record BarkNumber(double value) implements BarkValue {
    @Override
    public String display() {
      boolean isWholeNumber = Math.floor(value) == value && !Double.isInfinite(value);
      return isWholeNumber ? String.valueOf((long) value) : String.valueOf(value);
    }

    @Override
    public boolean equalTo(final BarkValue other) {
      return other instanceof BarkNumber(double v) && value == v;
    }
  }

  record BarkBoolean(boolean value) implements BarkValue {
    @Override
    public String display() {
      return value ? "true" : "false";
    }

    @Override
    public boolean equalTo(final BarkValue other) {
      return other instanceof BarkBoolean(boolean v) && value == v;
    }
  }

  record BarkNull() implements BarkValue {
    @Override
    public String display() {
      return "null";
    }

    @Override
    public boolean equalTo(final BarkValue other) {
      return other instanceof BarkNull;
    }
  }

  static BarkValue of(final String value) {
    return new BarkString(value);
  }

  static BarkValue of(final double value) {
    return new BarkNumber(value);
  }

  static BarkValue of(final boolean value) {
    return new BarkBoolean(value);
  }
}

File: pbark/pbark/interpreter/bark_value.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Union


@dataclass(frozen=True)
class BarkString:
    value: str

    def display(self) -> str:
        return self.value

    def equal_to(self, other: BarkValue) -> bool:
        return isinstance(other, BarkString) and self.value == other.value


@dataclass(frozen=True)
class BarkNumber:
    value: float

    def display(self) -> str:
        is_whole = self.value == int(self.value) and abs(self.value) != float("inf")
        return str(int(self.value)) if is_whole else str(self.value)

    def equal_to(self, other: BarkValue) -> bool:
        return isinstance(other, BarkNumber) and self.value == other.value


@dataclass(frozen=True)
class BarkBoolean:
    value: bool

    def display(self) -> str:
        return "true" if self.value else "false"

    def equal_to(self, other: BarkValue) -> bool:
        return isinstance(other, BarkBoolean) and self.value == other.value


@dataclass(frozen=True)
class BarkNull:
    def display(self) -> str:
        return "null"

    def equal_to(self, other: BarkValue) -> bool:
        return isinstance(other, BarkNull)


BarkValue = Union[BarkString, BarkNumber, BarkBoolean, BarkNull]


def describe(value: BarkValue | None) -> str:
    return "nothing" if value is None else value.display()


def of(value: str | float | bool) -> BarkValue:
    if isinstance(value, bool):
        return BarkBoolean(value)
    if isinstance(value, (int, float)):
        return BarkNumber(float(value))
    return BarkString(value)
